### PR TITLE
Update engine to use new results

### PR DESCRIPTION
**Pull request by madkinsz**
Hooks everything up! Everywhere in the engine where we were creating `DataDocument` types, we are now creating  new result types. 

The largest changes here are dealing with cases where we previously would pickle exceptions and attach them to states. We do this all over the place — sometimes where we do not have access to the user's result settings. To support those cases, I've refactored failed/crashed results to allow generation of an exception from a string. This means we do not need to actually serialize an exception to report a failure, e.g. when an agent fails to submit a flow run.

Unlike the other pull requests in #6908 this one is a little sprawling — at some point everything's got to switch over and little edge cases need to be addressed. I think this makes this a good candidate for synchronous review.

I have not added tests here covering usage of results. Instead, I've focused on fixing the existing tests which make extensive usage of results anyway. I'm going to add dedicated tests in a follow-up pull request but want to get this merged ASAP to facilitate QA of this feature.

### CODE REVIEW

**Review by abrookins — APPROVED** [review]
Wow, there's a lot here! I don't think I can give a detailed review, but I read through parts and never felt shocked. 😅 I'm more interested in getting this merged and into some users' hands to try it out!

File: src/prefect/client/schemas.py [hunk 1]
```diff
@@ -134,25 +135,20 @@ def result(self, raise_on_failure: bool = True, fetch: Optional[bool] = None):
...
-            return result_from_state_with_data_document(
-                self, raise_on_failure=raise_on_failure
-            )
-        elif isinstance(self.data, BaseResult):
-            return await self.data.get()
-        else:
-            raise ValueError(
-                f"State data is of unknown result type {type(self.data).__name__!r}."
-            )
+    def to_state_create(self) -> schemas.actions.StateCreate:
```
Comment (madkinsz) [inline src/prefect/client/schemas.py:142]: Add docstring

File: src/prefect/engine.py [hunk 2]
```diff
@@ -1096,6 +1063,8 @@ async def begin_task_run(
                 anyio.create_task_group()
             )
 
+        await stack.enter_async_context(report_task_run_crashes(task_run, client))
```
Comment (madkinsz) [inline src/prefect/engine.py:1066]: Investigate if this change can be split out

File: src/prefect/engine.py [hunk 3]
```diff
@@ -1319,24 +1257,30 @@ async def wait_for_task_runs_and_report_crashes(
         if not state.type == StateType.CRASHED:
             continue
 
-        exception = await state.result(raise_on_failure=False, fetch=True)
+        exception = await get_state_exception(state)
```
Comment (madkinsz) [inline src/prefect/engine.py:1260]: Figure out why we call this here instead of `state.result`

File: src/prefect/exceptions.py [hunk 4]
```diff
@@ -56,6 +56,22 @@ class PrefectException(Exception):
...
+    """
+
+
+class FailedRun(PrefectException):
+    """
+    Raised when the result from a failed run is retrieved and an exception is not
+    attached.
+
+    This occurs when a string is attached to the state instead of an exception
+    or the attached exception is a `BaseException` and is not safe to raise.
```
Comment (madkinsz) [inline src/prefect/exceptions.py:71]: Check if removed?

File: src/prefect/states.py [hunk 5]
```diff
@@ -121,51 +174,124 @@ async def return_value_to_state(result: Any, serializer: str = "cloudpickle") ->
...
     else:
         raise TypeError(
-            f"Unexpected result for failure state: {result!r} —— "
+            f"Unexpected result for failed state: {result!r} —— "
             f"{type(result).__name__} cannot be resolved into an exception"
         )
 
 
+@sync_compatible
+async def raise_state_exception(state: State) -> None:
```
Comment (madkinsz) [inline src/prefect/states.py:285]: Determine if this is used?

File: src/prefect/engine.py (hunk 2 above)
Comment (madkinsz) [reply src/prefect/engine.py:1066]: No. This is needed because now we raise base exceptions instead of returning them.
